WordPress/gutenberg
# Try Ariakit Select for new CustomSelectControl component

>[!NOTE]
>WIP: This component is experimental and will have many upcoming iterations. Please feel free to add any suggestions to the above tracking issue. 

## What?


Part of #55023

Try creating a new `CustomSelectControl` using Ariakit to include new features, including an option to render children. 

## Why?


For more customization potential and a more intuitive API, allowing the use of JSX to add select values rather than a strict object like in our current `CustomSelectControl` component. 

## How?

 
By creating a compound component: 

```jsx
<CustomSelect>
	<CustomSelectItem value="Small">
		<span style={ { fontSize: '75%' } }>Small</span>
	</CustomSelectItem>
	<CustomSelectItem value="Something bigger">
		<span style={ { fontSize: '200%' } }>Something bigger</span>
	</CustomSelectItem>
</CustomSelect>
```

## New Features

### Multi-Select

Option to select more than one value:

<img width="489" alt="Screenshot 2023-11-20 at 11 05 59 AM" src="https://github.com/WordPress/gutenberg/assets/35543432/5c43a107-d2d0-452f-8e03-5d8018143db2">

To render any JSX children, with rich previews: 

<img width="486" alt="Screenshot 2023-11-20 at 11 06 14 AM" src="https://github.com/WordPress/gutenberg/assets/35543432/d171abdb-22eb-479e-9b09-9a3d73f88024">

## Testing Instructions

1. `npm run storybook: dev`
2. Test out the component and ensure it works as expected as a select, with the new features mentioned above

Generally:

- Styles should match existing `CustomSelectControl` as closely as possible 
- Selecting a value changes the input

Per Story: 

- `Default`: Padding, font-size, and height should change for size `small` 
- `Multi Select`: Should be able to select and deselect multiple items and count in input should reflect this

## Follow-ups: 

- [Look into roles to see what is best for this component (i.e. combobox vs listbox)](https://github.com/WordPress/gutenberg/pull/55790#pullrequestreview-1738002817) 
- [Consider using `WordPressComponentProps`](https://github.com/WordPress/gutenberg/pull/55790#discussion_r1397930163)
- [Add `compact` size variant](https://github.com/WordPress/gutenberg/pull/55790#discussion_r1397834798) 
- [Look into selectors when polishing styles](https://github.com/WordPress/gutenberg/pull/55790#discussion_r1397953698)
- [Add a check to make sure children are `CustomSelectItem` components](https://github.com/WordPress/gutenberg/pull/55790#discussion_r1397956294)
- [Format README](https://github.com/WordPress/gutenberg/pull/55790#discussion_r1397931036)
